go/vt/mysqlctl/builtinbackupengine.go

```diff
@@ -517,66 +516,39 @@ func (be *BuiltinBackupEngine) ExecuteRestore(
 	restoreConcurrency int,
 	hookExtraEnv map[string]string) (mysql.Position, error) {
 
-	var bh backupstorage.BackupHandle
+	zeroPosition := mysql.Position{}
 	var bm builtinBackupManifest
-	var toRestore int
 
-	for toRestore = len(bhs) - 1; toRestore >= 0; toRestore-- {
-		bh = bhs[toRestore]
-		rc, err := bh.ReadFile(ctx, backupManifest)
-		if err != nil {
-			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), dir, err)
-			continue
-		}
-
-		err = json.NewDecoder(rc).Decode(&bm)
-		rc.Close()
-		if err != nil {
-			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage (cannot JSON decode MANIFEST: %v)", bh.Name(), dir, err)
-			continue
-		}
-
-		logger.Infof("Restore: found backup %v %v to restore with %v files", bh.Directory(), bh.Name(), len(bm.FileEntries))
-		break
+	bh, err := findBackupToRestore(ctx, cnf, mysqld, logger, dir, bhs, &bm)
+	if err != nil {
+		return zeroPosition, err
 	}
-	if toRestore < 0 {
-		// There is at least one attempted backup, but none could be read.
-		// This implies there is data we ought to have, so it's not safe to start
-		// up empty.
-		return mysql.Position{}, errors.New("backup(s) found but none could be read, unsafe to start up empty, restart to retry restore")
+	if err = prepareToRestore(ctx, cnf, mysqld, logger); err != nil {
+		return zeroPosition, err
 	}
 
-	// Starting from here we won't be able to recover if we get stopped by a cancelled
-	// context. Thus we use the background context to get through to the finish.
+	logger.Infof("Restore: copying %v files", len(bm.FileEntries))
 
-	logger.Infof("Restore: shutdown mysqld")
-	err := mysqld.Shutdown(context.Background(), cnf, true)
-	if err != nil {
-		return mysql.Position{}, err
+	if err = createStateFile(cnf); err != nil {
```

Comment:
Should we create the state file before `prepareToRestore()`? If `prepareToRestore()` fails at some point in the middle, couldn't it leave mysqld unable to start, causing the type of stuck state we're trying to prevent?

Reply:
I was wondering about this too. I think to be safe we should create if before prepareToRestore especially since that function deletes the existing db files.